### PR TITLE
Add time grain blacklist and addons to config.py

### PR DESCRIPTION
**Pull request by villebro**
- Refactor time grain logic to avoid the use of conflicting names and durations
- Harmonize discrepancies in time grains, e.g. P3M (Druid) vs P0.25Y (all others)
- Add TIME_GRAIN_BLACKLIST to config.py to be able to remove unwanted time grains, e.g. 5/10/15 mins
 - Add TIME_GRAIN_ADDONS and TIME_GRAIN_ADDON_FUNCTIONS to config.py to add custom time grains not natively supported by Superset
- Add tests for blacklist and validity of defined time grains

### CODE REVIEW

**Comment by mistercrunch** [comment]
@betodealmeida

**Comment by villebro** [comment]
Also added functionality to be able to add new custom time grains in `config.py` (`TIME_GRAIN_ADDONS` and `TIME_GRAIN_ADDON_FUNCTIONS`). This could be used to move some less used time grains out of db_engine_specs, while still allowing for end users to define their own custom time grains where needed.

Question: in `config.py`, which would be preferable: using a blacklist to remove predefined options (e.g. `VIZ_TYPE_BLACKLIST`), or effectively defining a whitelist (e.g. `LANGUAGES`)?

**Comment by villebro** [comment]
Comments would be much appreciated @betodealmeida @mistercrunch . This PR would make it easier and to manage and add new time grains in production environments. If this PR is not aligned with future plans for Superset I will be happy make necessary changes or close the PR.

**Review by betodealmeida — CHANGES_REQUESTED** [review]
This looks great, @villebro! It greatly simplifies the definition of new durations and reduces a lot of the duplication.

I have just a few comments on the code, mostly nits, and there's one place where the function is wrong (probably due to an accidental deletion).

> Question: in config.py, which would be preferable: using a blacklist to remove predefined options (e.g. VIZ_TYPE_BLACKLIST), or effectively defining a whitelist (e.g. LANGUAGES)?

I think doing a blacklist works better here.

File: superset/db_engine_specs.py [hunk 1]
```diff
@@ -52,6 +52,35 @@
...
+    'PT0.5H': 'half hour',
+    'PT1H': 'hour',
+    'P1D': 'day',
+    'P1W': 'week',
+    'P1M': 'month',
+    'P0.25Y': 'quarter',
+    'P1Y': 'year',
+    '1969-12-28T00:00:00Z/P1W': 'week_start_sunday',
+    '1969-12-29T00:00:00Z/P1W': 'week_start_monday',
+    'P1W/1970-01-03T00:00:00Z': 'week_ending_saturday',
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Good idea, done.

File: superset/db_engine_specs.py [hunk 2]
```diff
@@ -52,6 +52,35 @@
...
+}
+
+
+def _create_time_grains_tuple(time_grains, time_grain_functions, blacklist):
+    ret_list = list()
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Thanks, will try to remember this in the future.

File: superset/db_engine_specs.py [hunk 3]
```diff
@@ -52,6 +52,35 @@
...
+    blacklist = blacklist if blacklist else []
+    for duration, func in time_grain_functions.items():
+        if duration not in blacklist:
+            name = time_grains.get(duration, None)
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Good point.

File: superset/db_engine_specs.py [hunk 4]
```diff
@@ -65,12 +94,22 @@ class BaseEngineSpec(object):
     """Abstract class for database engine specific configurations"""
 
     engine = 'base'  # str as defined in sqlalchemy.engine.engine
-    time_grains = tuple()
+    time_grain_functions = dict()
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Fixed.

File: superset/db_engine_specs.py [hunk 5]
```diff
@@ -52,6 +52,35 @@
 
 Grain = namedtuple('Grain', 'name label function duration')
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Good point, I misremembered that `core.grains_dict()` created a dict that mapped `duration -> grain` and `name -> grain`, but the latter was in fact `label -> grain`. I've now removed name.

File: superset/db_engine_specs.py [hunk 6]
```diff
@@ -65,12 +94,22 @@ class BaseEngineSpec(object):
...
 
+    @classmethod
+    def time_grains(cls):
+        blacklist = config.get('TIME_GRAIN_BLACKLIST', [])
+        grains = builtin_time_grains.copy()
+        grains.update(config.get('TIME_GRAIN_ADDONS', {}))
+        grain_functions = cls.time_grain_functions.copy()
+        grain_addon_functions = config.get('TIME_GRAIN_ADDON_FUNCTIONS', {})
+        grain_functions.update(grain_addon_functions.get(cls.engine, {}))
+        return _create_time_grains_tuple(grains, grain_functions, blacklist)
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Changed.

File: superset/db_engine_specs.py [hunk 7]
```diff
@@ -340,25 +379,17 @@ class PostgresBaseEngineSpec(BaseEngineSpec):
...
-        Grain('month', _('month'),
-              "DATE_TRUNC('month', {col}) AT TIME ZONE 'UTC'", 'P1M'),
-        Grain('quarter', _('quarter'),
-              "DATE_TRUNC('quarter', {col}) AT TIME ZONE 'UTC'", 'P0.25Y'),
-        Grain('year', _('year'),
-              "DATE_TRUNC('year', {col}) AT TIME ZONE 'UTC'", 'P1Y'),
-    )
+    time_grain_functions = {
+        None: '{col}',
+        'PT1S': "{col}) AT TIME ZONE 'UTC'",
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Whoops, good catch!

File: superset/db_engine_specs.py [hunk 8]
```diff
@@ -390,17 +421,25 @@ def get_table_names(cls, schema, inspector):
...
-        Grain('week', _('week'), "DATE_TRUNC('WEEK', {col})", 'P1W'),
-        Grain('month', _('month'), "DATE_TRUNC('MONTH', {col})", 'P1M'),
-        Grain('quarter', _('quarter'), "DATE_TRUNC('QUARTER', {col})", 'P0.25Y'),
-        Grain('year', _('year'), "DATE_TRUNC('YEAR', {col})", 'P1Y'),
-    )
+    time_grain_functions = {
+        None: '{col}',
+        'PT1S': "DATE_TRUNC('SECOND', {col})",
+        'PT1M': "DATE_TRUNC('MINUTE', {col})",
+        'PT5M': "DATEADD(MINUTE, FLOOR(DATE_PART(MINUTE, {col}) / 5) * 5, \
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: Yes; not sure if it's the most efficient way of doing it, but managed to get good performance on a semi-large dataset, so should be good to go.

File: superset/db_engine_specs.py [hunk 9]
```diff
@@ -340,25 +379,17 @@ class PostgresBaseEngineSpec(BaseEngineSpec):
...
+        None: '{col}',
+        'PT1S': "{col}) AT TIME ZONE 'UTC'",
+        'PT1M': "DATE_TRUNC('minute', {col}) AT TIME ZONE 'UTC'",
+        'PT1H': "DATE_TRUNC('hour', {col}) AT TIME ZONE 'UTC'",
+        'P1D': "DATE_TRUNC('day', {col}) AT TIME ZONE 'UTC'",
+        'P1W': "DATE_TRUNC('week', {col}) AT TIME ZONE 'UTC'",
+        'P1M': "DATE_TRUNC('month', {col}) AT TIME ZONE 'UTC'",
+        'P0.25Y': "DATE_TRUNC('quarter', {col}) AT TIME ZONE 'UTC'",
+        'P1Y': "DATE_TRUNC('year', {col}) AT TIME ZONE 'UTC'",
+    }
```
[the remark this answers is not in the document]
Comment (villebro) [reply superset/db_engine_specs.py]: 👍

**Review by villebro — COMMENTED** [review]
All proposed changes implemented.

**Comment by villebro** [comment]
@betodealmeida I tried to remove `name` from Grain and updating `time_column_grains()` and `data()` in `connectors/sqla/models.py`, but got slightly overwhelmed by the frontend code. For instance, `controls.jsx` seems to have similar but slightly different grains, and am unsure where these definitions are used:

https://github.com/apache/incubator-superset/blob/bea0a0aa1590d073365ecf74d75a054643dbe3b7/superset/assets/src/explore/controls.jsx#L750-L776

I'm sure I will be able to sort out how the frontend logic works, but it will take time, so I propose merging this and removing `name` in a separate PR.

**Review by betodealmeida — APPROVED** [review]
Awesome! Thank you so much for doing this, @villebro!

**Comment by villebro** [comment]
Great, thanks @betodealmeida !

**Comment by villebro** [comment]
@betodealmeida FYI rebased to resolve conflict, any chance of getting this merged out of the way?

**Comment by betodealmeida** [comment]
@mistercrunch, can you merge this? I still don't have merging privilege.

**Pull request merged**